ch01python/00pythons.ipynb.py

```diff
@@ -12,224 +12,241 @@
+# computation being displayed before returning to the prompt again.
+#
+
+# %% [markdown]
+# ### Python libraries
+#
+# A very common requirement in research (and all other!) programming is needing to reuse code in
+# multiple different files. While it may seem that copying-and-pasting is an adequate solution to this
+# problem, this should generally be avoided wherever possible and code which we wish to reuse
+# _factored out_ in to _libraries_ which we we can _import_ in to other files to access the functionality
+# of this code. 
+#
+# Compared to copying and pasting code, writing and using libraries has the major advantage of meaning
+# if we spot a bug in the code we only need to fix it once in the underlying library, and we straight away
+# have the fixed code available everywhere the library is used rather than having to separately implement
+# the fix in each file it is used. This similarly applies to for example adding new features to a piece of
+# code. By creating libraries we also make it easier for other researchers to use our code.
+#
+# While it is simple to use libraries within a notebook (and we have already seen examples of this when we
+# imported the Python libraries NumPy and Matplotlib in the figure-eight plot example above), it is non-trivial
+# to use code from one notebook in another without copying-and-pasting. To create Python libraries we therefore
+# generally write the code in to text files with a `.py` extension which in Python terminology are called _modules_.
+# The code can in these file can then be used in notebooks (or other modules) using the Python `import` statement.
+# For example the cell below creates a file `draw_eight.py` in the same directory as this notebook containing Python
+# code defining a _function_ (we will cover how to define and call functions later in the course) which creates a
+# figure-eight plot and return the figure object.
 
 # %%
-# %%writefile draw_eight.py 
-# Above line tells the notebook to treat the rest of this
-# cell as content for a file on disk.
-import math
+# %%writefile draw_eight.py
+# The above line tells the notebook to write the rest of the cell content to a file draw_eight.py
 
 import numpy as np
 import matplotlib.pyplot as plt
 
 def make_figure():
-    """Plot a figure of eight."""
-
-    theta = np.arange(0, 4 * math.pi, 0.1)
-    eight = plt.figure()
-    axes = eight.add_axes([0, 0, 1, 1])
-    axes.plot(0.5 * np.sin(theta), np.cos(theta / 2))
-
-    return eight
-
+    theta = np.linspace(0, 2 * np.pi, 100)
+    fig, ax = plt.subplots(figsize=(3, 6))
+    ax.plot(np.sin(2 * theta) / 2, np.cos(theta))
+    return fig
 
 
 # %% [markdown]
-# In a real example, we could edit the file on disk
-# using a code editor such as [VS code](https://code.visualstudio.com/).
+# We can use this code in the notebook by _importing_ the `draw_eight` module and then _calling_ the
```

Comment:
```suggestion
# Note, in a real example, we could edit the file on disk using a code editor (or IDE) rather than using `%%writefile`
#
# We can use this code in the notebook by _importing_ the `draw_eight` module and then _calling_ the
```